### PR TITLE
TileCoord supports up to zoom 15 using alternate ordering

### PR DESCRIPTION
**Pull request by bdon**
This is a rough idea right now (don't merge yet) to explore alternate TileCoord encodings. it deserves some better tests for edge cases as right now it's blindly ported from C++ to java

The current TileCoord uses a fixed 14 bits for X and Y and the remaining bits for Z. This instead defines a mapping from a 32-bit unsigned encoded integer to X,Y,Z like this:

Tile 0 is `(0,0)` on z0
Tile 1-4 are positions on the z1 Hilbert curve; each side of the curve is length 2, so tile 1 is `(0,0)`, tile 2 is `(0,1)`, tile 3 is `(1,1)` and tile 4 is `(1,0)`
Tile 5-20 are positions on the z2 Hilbert curve
etc...

Drawbacks:
* decoding/encoding might more expensive
* does not correspond to index insertion order in MBTiles, but there may be other ways to mitigate this

Potential benefits of doing it this way:
* support z15 without widening the TileCoord encoding to 64 bits
* If features are ordered this way and memory-mapped on disk, reads should have better locality for 2D queries vs. striped Z/X/Y ordering
* for some other tools I'm developing this would aid interoperability, such as a bitmap format for sparse ocean representation or a future ordering requirement for PMTiles

### CODE REVIEW

**Comment by msbarry** [comment]
Thanks for implementing this!  A couple of concerns I'll want to mitigate before merging this:

1. Any areas within planetiler that assume tiles are ordered by z asc, x asc, y desc - I think there's a few but will need to dig a bit more - the one testSkipFill failing test might be indicative of something larger not covered by other tests.
2. Test performance impact for tile generation - we _could_ write a microbenchmark for TileCoord encode/decode but I'd probably rather just test generating on a large extract or the planet and look at overall difference in performance.
3. Test difference in mbtiles read performance - you can give two mbtiles files to `BenchmarkMbtilesRead` and it reports the qps difference between the two, but reads tiles randomly - I assume we'd need something different to simulate more realistic traffic?

Also, do you have a link to the C++ code you ported? Want to check if there's any license/attribution requirements.

**Comment by msbarry** [comment]
Also for the "Check formatting" failure - if you run `./scripts/format.sh` to autoformat the code and push that result it should fix the issue.  I've got some instructions in [CONTRIBUTING](https://github.com/onthegomap/planetiler/blob/main/CONTRIBUTING.md#ide-setup) for setting up autoformat in a couple of IDEs if you were interested in that as well.

**Comment by bdon** [comment]
Thanks, I'll get started on these tasks. The implementation is part of my own C++ library which is not on GitHub.

**Comment by msbarry** [comment]
Took a quick look for potential issues I can see with tiles not ordered by z/x/y:

- the previous ordering matches the sqlite index order exactly, it seemed like that would be optimal for building the index (size and speed) - to test that theory, could you post timing and size summaries printed at the end or a planet run for old z/x/y ordering and hilbert ordering?
-  `MbtilesWriter.tileEncoder` skips re-encoding/computing some things if the previous tile is the same as the current tile - for z/x/y this mean processing columns of filled polygons sequentially.  For hilbert ordering that should be fine too since sequential tiles are adjacent?
- Progress logging during mbtiles writes expects we process all z0 tiles, then z1, then z2 - that should still be the case for hilbert ordering?
- `MbtilesWriter.getLastTileLogDetails` logs progress during mbtiles writes prints % complete for each zoom level, and a osm link to what part of the world it's working on so you can debug if things are slow.  The OSM link should be more helpful with hilbert ordering since it's working on a part of the world instead of vertical slice.  Is there a way to get a rough idea of % complete within a zoom level though? The previous logic used `(100 * (lastTile.x() + 1 - zMinX)) / (zMaxX - zMinX)` since we work in order across X columns.

Nothing before tile writing _should_ care about tile ordering.

File: planetiler-core/src/test/java/com/onthegomap/planetiler/PlanetilerTests.java [hunk 1]
```diff
@@ -687,7 +687,7 @@ void testSkipFill() throws Exception {
         .setBufferPixels(4)
     );
 
-    assertEquals(481, results.tiles.size());
+    assertEquals(130, results.tiles.size());
```
Comment (msbarry) [inline planetiler-core/src/test/java/com/onthegomap/planetiler/PlanetilerTests.java:690]: @bdon this was fixed by #279 - you can revert this and rebase and it should be fine now.

**Comment by msbarry** [comment]
Re: performance:

- for encoding, it looks like there are several places that construct `TileCoord` instances just to store features in a map but don't actually care about the encoded int value.  Only `FeatureGroup.encodeKey` accesses the encoded value outside of tests
- for decoding, we only decode the value from `FeatureGroup.TileFeatures` constructor once per output tile

So it might make sense to just compute the encoded value on-demand instead of storing on the `TileCoord` instance, or computing it lazily now that it's not "free"

File: planetiler-core/src/main/java/com/onthegomap/planetiler/geo/TileCoord.java [hunk 2]
```diff
@@ -92,7 +92,7 @@ private static long hilbertIndexToXY(int n, int i) {
   }
 
   // Ignore warnings about nested scopes amd code duplication.
-  @SuppressWarnings({"java:S1199", "common-java:DuplicatedBlocksSource"})
+  @SuppressWarnings({"java:S1199", "common-java:DuplicatedBlocks"})
```
Comment (bdon) [inline planetiler-core/src/main/java/com/onthegomap/planetiler/geo/TileCoord.java:95]: this is not pretty but IMHO we should just leave the nested scope formatting to preserve the original code structure

**Comment by bdon** [comment]
run of `planetiler.jar` (Jun 5 release) 16cores

```
java -Xmx60g -jar planetiler.jar --area=planet --bounds=planet --download --download-threads=10 --download-chunk-size-mb=1000 --fetch-wikidata --mbtiles=output.mbtiles --nodemap-type=array --storage=mmap --force
```

```
2:56:05 DEB [mbtiles] - Tile stats:
2:56:05 DEB [mbtiles] - z0 avg:76k max:76k
2:56:05 DEB [mbtiles] - z1 avg:172k max:194k
2:56:05 DEB [mbtiles] - z2 avg:680k max:1.1M
2:56:05 DEB [mbtiles] - z3 avg:221k max:1M
2:56:05 DEB [mbtiles] - z4 avg:79k max:744k
2:56:05 DEB [mbtiles] - z5 avg:40k max:628k
2:56:05 DEB [mbtiles] - z6 avg:18k max:421k
2:56:05 DEB [mbtiles] - z7 avg:17k max:764k
2:56:05 DEB [mbtiles] - z8 avg:9.9k max:579k
2:56:05 DEB [mbtiles] - z9 avg:6.6k max:502k
2:56:05 DEB [mbtiles] - z10 avg:2.9k max:323k
2:56:05 DEB [mbtiles] - z11 avg:1.4k max:188k
2:56:05 DEB [mbtiles] - z12 avg:772 max:239k
2:56:05 DEB [mbtiles] - z13 avg:411 max:267k
2:56:05 DEB [mbtiles] - z14 avg:366 max:1.8M
2:56:05 DEB [mbtiles] - all avg:425 max:0
2:56:05 DEB [mbtiles] -  # features: 2,955,775,444
2:56:05 DEB [mbtiles] -     # tiles: 267,718,292
2:56:05 INF [mbtiles] - Finished in 1h17m9s cpu:18h15m38s gc:6m14s avg:14.2
2:56:05 INF [mbtiles] -   read    1x(12% 9m24s sys:2m36s wait:56m7s done:4s)
2:56:05 INF [mbtiles] -   encode 15x(86% 1h6m36s sys:15s wait:3m38s done:4s)
2:56:05 INF [mbtiles] -   write   1x(16% 12m11s sys:1m58s wait:58m19s)
2:56:06 INF - Finished in 2h56m6s cpu:38h56m49s gc:11m54s avg:13.3
2:56:06 INF - FINISHED!
2:56:06 INF -
2:56:06 INF - ----------------------------------------
2:56:06 INF -   overall          2h56m6s cpu:38h56m49s gc:11m54s avg:13.3
2:56:06 INF -   wikidata         5m14s cpu:1h18m42s gc:7s avg:15.1
2:56:06 INF -     pbf     1x(70% 3m39s sys:2m11s wait:1m15s done:9s)
2:56:06 INF -     filter 15x(93% 4m51s sys:1s wait:5s done:9s)
2:56:06 INF -     fetch   1x(1% 4s wait:5m2s done:8s)
2:56:06 INF -   lake_centerlines 2s cpu:7s avg:3.8
2:56:06 INF -     read     1x(84% 1s)
2:56:06 INF -     process 15x(5% 0.1s wait:2s)
2:56:06 INF -     write    1x(0% 0s wait:2s)
2:56:06 INF -   water_polygons   1m10s cpu:16m5s gc:3s avg:13.7
2:56:06 INF -     read     1x(22% 15s wait:48s done:2s)
2:56:06 INF -     process 15x(82% 57s wait:8s)
2:56:06 INF -     write    1x(37% 26s sys:8s wait:43s)
2:56:06 INF -   natural_earth    5s cpu:18s avg:3.3
2:56:06 INF -     read     1x(91% 5s)
2:56:06 INF -     process 15x(9% 0.5s wait:5s)
2:56:06 INF -     write    1x(0% 0s wait:5s)
2:56:06 INF -   osm_pass1        5m47s cpu:1h24m18s gc:8s avg:14.6
2:56:06 INF -     read     1x(49% 2m49s sys:1m42s wait:2m16s done:2s)
2:56:06 INF -     process 15x(91% 5m18s sys:8s block:6s wait:12s)
2:56:06 INF -   osm_pass2        1h13m21s cpu:15h48m56s gc:3m9s avg:12.9
2:56:06 INF -     read     1x(3% 2m14s sys:1m28s wait:1h10m14s done:36s)
2:56:06 INF -     process 15x(80% 58m52s sys:1m47s block:6s wait:6s)
2:56:06 INF -     write    1x(27% 19m59s sys:4m1s wait:51m37s)
2:56:06 INF -   boundaries       10s cpu:15s avg:1.4
2:56:06 INF -   sort             12m53s cpu:1h52m9s gc:2m11s avg:8.7
2:56:06 INF -     worker 16x(19% 2m27s sys:14s wait:2m11s done:13s)
2:56:06 INF -   mbtiles          1h17m9s cpu:18h15m38s gc:6m14s avg:14.2
2:56:06 INF -     read    1x(12% 9m24s sys:2m36s wait:56m7s done:4s)
2:56:06 INF -     encode 15x(86% 1h6m36s sys:15s wait:3m38s done:4s)
2:56:06 INF -     write   1x(16% 12m11s sys:1m58s wait:58m19s)
2:56:06 INF - ----------------------------------------
2:56:06 INF -   features        208GB
2:56:06 INF -   mbtiles 81GB
```

**Comment by bdon** [comment]
Branch planet run output:

```
2:57:24 DEB [mbtiles] - Tile stats:
2:57:24 DEB [mbtiles] - z0 avg:76k max:76k
2:57:24 DEB [mbtiles] - z1 avg:172k max:194k
2:57:24 DEB [mbtiles] - z2 avg:680k max:1.1M
2:57:24 DEB [mbtiles] - z3 avg:221k max:1M
2:57:24 DEB [mbtiles] - z4 avg:79k max:744k
2:57:24 DEB [mbtiles] - z5 avg:40k max:628k
2:57:24 DEB [mbtiles] - z6 avg:18k max:421k
2:57:24 DEB [mbtiles] - z7 avg:17k max:764k
2:57:24 DEB [mbtiles] - z8 avg:9.9k max:579k
2:57:24 DEB [mbtiles] - z9 avg:6.6k max:502k
2:57:24 DEB [mbtiles] - z10 avg:2.9k max:323k
2:57:24 DEB [mbtiles] - z11 avg:1.4k max:188k
2:57:24 DEB [mbtiles] - z12 avg:772 max:239k
2:57:24 DEB [mbtiles] - z13 avg:411 max:267k
2:57:24 DEB [mbtiles] - z14 avg:366 max:1.8M
2:57:24 DEB [mbtiles] - all avg:425 max:0
2:57:24 DEB [mbtiles] -  # features: 2,955,775,550
2:57:24 DEB [mbtiles] -     # tiles: 267,718,292
2:57:24 INF [mbtiles] - Finished in 1h17m19s cpu:18h23m10s gc:5m33s avg:14.3
2:57:24 INF [mbtiles] -   read    1x(13% 9m41s sys:2m39s wait:56m34s done:5s)
2:57:24 INF [mbtiles] -   encode 15x(86% 1h6m37s sys:22s wait:4m6s done:5s)
2:57:24 INF [mbtiles] -   write   1x(18% 13m56s sys:2m46s wait:59m18s)
2:57:24 INF - Finished in 2h57m24s cpu:38h58m1s gc:10m56s avg:13.2
2:57:24 INF - FINISHED!
2:57:24 INF -
2:57:24 INF - ----------------------------------------
2:57:24 INF -   overall          2h57m24s cpu:38h58m1s gc:10m56s avg:13.2
2:57:24 INF -   wikidata         5m16s cpu:1h18m16s gc:7s avg:14.8
2:57:24 INF -     pbf     1x(75% 3m57s sys:2m32s wait:57s done:9s)
2:57:24 INF -     filter 15x(91% 4m48s sys:1s wait:9s done:9s)
2:57:24 INF -     fetch   1x(1% 4s wait:5m5s done:8s)
2:57:24 INF -   lake_centerlines 2s cpu:6s avg:3.5
2:57:24 INF -     read     1x(83% 1s)
2:57:24 INF -     process 15x(4% 0.1s wait:2s)
2:57:24 INF -     write    1x(0% 0s wait:2s)
2:57:24 INF -   water_polygons   1m11s cpu:15m52s gc:3s avg:13.5
2:57:24 INF -     read     1x(20% 14s wait:50s done:3s)
2:57:24 INF -     process 15x(80% 56s wait:9s)
2:57:24 INF -     write    1x(38% 27s sys:8s wait:42s)
2:57:24 INF -   natural_earth    5s cpu:21s avg:4
2:57:24 INF -     read     1x(88% 5s)
2:57:24 INF -     process 15x(13% 0.7s wait:4s)
2:57:24 INF -     write    1x(0% 0s wait:5s)
2:57:24 INF -   osm_pass1        5m52s cpu:1h23m3s gc:9s avg:14.1
2:57:24 INF -     read     1x(53% 3m8s sys:2m1s wait:1m45s done:2s)
2:57:24 INF -     process 15x(88% 5m11s sys:9s block:6s wait:22s)
2:57:24 INF -   osm_pass2        1h14m21s cpu:15h51m gc:2m57s avg:12.8
2:57:24 INF -     read     1x(3% 2m28s sys:1m40s wait:1h10m53s done:37s)
2:57:24 INF -     process 15x(80% 59m7s sys:1m58s block:9s wait:9s)
2:57:24 INF -     write    1x(28% 20m46s sys:4m8s wait:51m49s)
2:57:24 INF -   boundaries       10s cpu:15s avg:1.4
2:57:24 INF -   sort             12m54s cpu:1h45m34s gc:2m5s avg:8.2
2:57:24 INF -     worker 16x(18% 2m16s sys:14s wait:2m33s done:15s)
2:57:24 INF -   mbtiles          1h17m19s cpu:18h23m10s gc:5m33s avg:14.3
2:57:24 INF -     read    1x(13% 9m41s sys:2m39s wait:56m34s done:5s)
2:57:24 INF -     encode 15x(86% 1h6m37s sys:22s wait:4m6s done:5s)
2:57:24 INF -     write   1x(18% 13m56s sys:2m46s wait:59m18s)
2:57:24 INF - ----------------------------------------
2:57:24 INF -   features        208GB
2:57:24 INF -   mbtiles 82GB
```

**Comment by bdon** [comment]
MBTiles benchmark for BRANCH:

```
worker@Ubuntu-2004-focal-64-minimal:~$ java -cp planetiler-dist-0.5-SNAPSHOT-with-deps.jar com.onthegomap.planetiler.benchmarks.BenchmarkMbtilesRead --bench_mbtiles=output.mbtiles
0:00:00 DEB - argument: bench_repetitions=10 (number of repetitions)
0:00:00 DEB - argument: bench_nr_tile_reads=500000 (number of tiles to read)
0:00:00 DEB - argument: bench_pre_warms=3 (number of pre warm runs)
0:00:00 DEB - argument: bench_mbtiles=output.mbtiles (the mbtiles file to read from)
0:08:41 INF - working on output.mbtiles
0:09:24 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1603208.694870, min=159937.217605, average=160320.869487, max=160890.394421}
0:09:24 INF - diffs
```

**Comment by bdon** [comment]
MBTiles benchmark for MAIN: 

```
java -cp planetiler-dist-0.5-SNAPSHOT-with-deps.jar com.onthegomap.planetiler.benchmarks.BenchmarkMbtilesRead --bench_mbtiles=output.mbtiles
0:00:00 DEB - argument: bench_repetitions=10 (number of repetitions)
0:00:00 DEB - argument: bench_nr_tile_reads=500000 (number of tiles to read)
0:00:00 DEB - argument: bench_pre_warms=3 (number of pre warm runs)
0:00:00 DEB - argument: bench_mbtiles=output.mbtiles (the mbtiles file to read from)
0:02:32 INF - working on output.mbtiles
0:03:14 INF - readOperationsPerSecondStats: DoubleSummaryStatistics{count=10, sum=1629460.846341, min=162661.918839, average=162946.084634, max=163245.337260}
0:03:14 INF - diffs
```

sqlite3_analyzer output for MAIN: 

https://gist.github.com/bdon/fe0111a676d8ba6bf432538cd6b69344

**Comment by msbarry** [comment]
Thanks @bdon, looks like hilbert ordering is a tiny bit slower on random reads but that might not be a very accurate test.  Seems like hilbert ordering is meant to help with more realistic traffic distribution on machine with less RAM.  What have you done in the past to test pmtiles reads with a more realistic traffic distribution?

**Comment by bdon** [comment]
I've only seen the gains from the ordering in serving latency from production traffic, but a reproducible benchmark would be good to add to https://github.com/bdon/TileSiege ; I think if we could get an anonymized day of real traffic that would be a more effective benchmark for testing query latency. It does look like the hilbert order also ends up with less full pages so the db ends up larger, but I wasn't running it with the vacuum DB option (is that the default?)
